File: LeetCode/LeetCode/matrix_search.go
package main

import "fmt"

func searchMatrix(mat [][]int, target int) bool {
	m := len(mat)
	if m == 0 {
		return false
	}

	n := len(mat[0])
	if n == 0 {
		return false
	}

	if target < mat[0][0] || target > mat[m-1][n-1]  {
		return false
	}

	r:=0
	for r < m && mat[r][0] <= target {
		r++
	}
	r--

	i, j :=0, n-1
	for i <= j {
		med := (i + j)/2
		switch {
		case mat[r][med] < target:
			i = med + 1
		case target < mat[r][med]:
			j = med -1
		default:
			return true
		}
	}
	return mat[r][j] == target
}

func main() {
	arr := [][]int {
		[]int{
			1,2,3,4,
		},
		[]int{
			5,6,7,8,
		},
		[]int{
			9,10,11,12,
		},
	}
	fmt.Println( searchMatrix(arr, 8))
	fmt.Println( searchMatrix(arr, 13))
}

File: LeetCode/LeetCode/matrix_search.cc
#include<iostream>
#include<vector>

using std::vector;

class Solution {
    public:
        bool searchMatrix(vector<vector<int>> &matrix, int target) {
            if (matrix.empty()) {
                return false;
            }

            const int  m = matrix.size();
            const int n = matrix[0].size();
            int left = 0;
            int right = m * n - 1;

            while( left <= right) {
                int mid = left + (right - left)/2;
                if(matrix[mid /n][mid %n] >= target) {
                    right = mid - 1;
                }else {
                    left = mid + 1;
                }
            }

            if (left != m*n && matrix[left / n][left % n] == target) {
                return true;
            }

            return false;
        }
};

int main() {
    Solution s;
    vector<vector<int>> v = {
        {
            1,2,3,4,5
        },
        {
            6,7,8,9,10
        },
        {
            11,12,13,14,15
        }
    };

    std::cout<<s.searchMatrix(v, 99)<<std::endl;
    std::cout<<s.searchMatrix(v, 8)<<std::endl;
    return 0;
}
